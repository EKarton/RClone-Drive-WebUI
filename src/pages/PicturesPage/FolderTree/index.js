import TreeView from '@mui/lab/TreeView';
import ExpandMoreIcon from '@mui/icons-material/ExpandMore';
import ChevronRightIcon from '@mui/icons-material/ChevronRight';
import FolderTreeItem from './FolderTreeItem';

export default function FolderTree({ remotes, onFolderSelect }) {
  const handleNodeSelect = (_e, nodeIds) => {
    onFolderSelect(nodeIds);
  };

  return (
    <TreeView
      aria-label="file tree navigator"
      defaultCollapseIcon={<ExpandMoreIcon />}
      defaultExpandIcon={<ChevronRightIcon />}
      onNodeSelect={handleNodeSelect}
    >
      {remotes.map((remote) => (
        <FolderTreeItem remote={remote} curPath="" label={remote} />
      ))}
    </TreeView>
  );
}
