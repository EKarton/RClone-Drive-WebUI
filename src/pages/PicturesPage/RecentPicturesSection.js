import Card from '@mui/material/Card';
import CardActionArea from '@mui/material/CardActionArea';
import Typography from '@mui/material/Typography';
import axios from 'axios';
import { useEffect, useState } from 'react';
import AutoSizer from 'react-virtualized-auto-sizer';
import Image from 'components/Image';
import useRCloneClient from 'hooks/rclone/useRCloneClient';
import useFileViewerDialog from 'hooks/utils/useFileViewerDialog';
import useRecentlyViewedImages from 'hooks/utils/useRecentlyViewedImages';
import getExistingPictures from 'utils/getExistingPictures';
import './RecentPicturesSection.scss';

/**
 * Represents the list of recently viewed images
 * in the Pictures page
 */
export default function RecentPicturesSection() {
  const { recentPictures, addImage } = useRecentlyViewedImages();
  const rCloneClient = useRCloneClient();
  const fileViewer = useFileViewerDialog();

  const [existingPictures, setExistingPictures] = useState([]);

  useEffect(() => {
    const cancelSource = axios.CancelToken.source();
    const cancelToken = cancelSource.token;

    getExistingPictures(recentPictures, rCloneClient, cancelToken).then((pictures) => {
      setExistingPictures(pictures);
    });

    return () => {
      cancelSource.cancel();
    };
  }, [rCloneClient, recentPictures]);

  if (existingPictures.length === 0) {
    return null;
  }

  const handleImageClicked = (image) => () => {
    addImage(image);
    fileViewer.show(image);
  };

  return (
    <>
      <Typography variant="h7" component="div" color="text.primary">
        <strong>Recently Viewed Pictures</strong>
      </Typography>
      <div className="recently-viewed-image">
        <AutoSizer>
          {({ height, width }) => {
            const numImagesToShow = width < 1920 ? 4 : 6;
            const imagesToShow = existingPictures.slice(0, numImagesToShow);

            const numFillers = Math.max(0, numImagesToShow - imagesToShow.length);

            return (
              <div style={{ width, height }} className="recently-viewed-image__wrapper">
                {imagesToShow.map((image) => (
                  <Card className="recently-viewed-image__card" key={image.fileName}>
                    <CardActionArea
                      className="recently-viewed-image__card-area"
                      onClick={handleImageClicked(image)}
                      data-testid={image.fileName}
                    >
                      <Image
                        image={image}
                        imgClassName="recently-viewed-image__img-image"
                        skeletonClassName="recently-viewed-image__img-skeleton"
                      />
                    </CardActionArea>
                  </Card>
                ))}
                {Array.from({ length: numFillers }, (_, i) => (
                  <div
                    key={i}
                    className="recently-viewed-image__card"
                    data-testid="image-fillers"
                  />
                ))}
              </div>
            );
          }}
        </AutoSizer>
      </div>
    </>
  );
}
