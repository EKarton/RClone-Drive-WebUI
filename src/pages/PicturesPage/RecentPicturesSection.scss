.recently-viewed-image {
  height: 9rem;
  width: 100%;

  &__wrapper {
    display: flex;
    justify-content: stretch;
    align-content: stretch;
    grid-gap: 1rem;
    overflow: clip;
  }

  &__card {
    flex-grow: 1;
  }

  &__card-area {
    width: 100%;
    height: 100%;
  }

  &__img-wrapper {
    height: 100%;
  }

  &__img-skeleton {
    height: 100% !important;
  }
}
