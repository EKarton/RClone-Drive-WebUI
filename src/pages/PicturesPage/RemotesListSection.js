import RemoteCardList from 'components/RemoteCardList';
import { StatusTypes } from 'utils/constants';
import useFetchRemotes from 'hooks/fetch-data/useFetchRemotes';

/**
 * Represents the list of remotes in the Pictures page
 */
export default function RemotesListSection({ onRemoteCardClicked }) {
  const { status, data: remotes } = useFetchRemotes();

  if (status === StatusTypes.LOADING) {
    return null;
  }

  if (status === StatusTypes.ERROR) {
    return 'Error!';
  }

  const handleRemoteCardClicked = (remote) => {
    onRemoteCardClicked(remote);
  };

  return (
    <>
      <h4>Your Remotes</h4>
      <RemoteCardList remotes={remotes} onClick={handleRemoteCardClicked} />
    </>
  );
}
