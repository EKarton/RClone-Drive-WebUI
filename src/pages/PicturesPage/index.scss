.pictures-page {
  padding-bottom: 3rem;
  display: flex;
  flex-direction: column;
  row-gap: 1rem;
}
