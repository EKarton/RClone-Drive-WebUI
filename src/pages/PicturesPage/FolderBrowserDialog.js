import Button from '@mui/material/Button';
import Dialog from '@mui/material/Dialog';
import DialogActions from '@mui/material/DialogActions';
import DialogContent from '@mui/material/DialogContent';
import DialogTitle from '@mui/material/DialogTitle';
import { useState } from 'react';
import FolderTree from 'components/FolderTree';

/**
 * Represents the dialog that appears when a user clicks on a remote
 */
export default function FolderBrowserDialog({ remotes, open, onCancel, onOk, title }) {
  const [remotePath, setRemotePath] = useState('');

  const handleFolderTreeSelected = (remotePath) => {
    setRemotePath(remotePath);
  };

  const handleOk = () => {
    if (!remotePath) {
      onCancel();
      return;
    }

    onOk(remotePath);
  };

  return (
    <Dialog
      sx={{ '& .MuiDialog-paper': { width: '80%', maxHeight: 435 } }}
      maxWidth="xs"
      open={open}
      onClose={onCancel}
    >
      <DialogTitle>{title}</DialogTitle>
      <DialogContent dividers>
        <FolderTree remotes={remotes} onSelect={handleFolderTreeSelected} />
      </DialogContent>
      <DialogActions>
        <Button autoFocus onClick={onCancel} data-testid="cancel-button">
          Cancel
        </Button>
        <Button onClick={handleOk} data-testid="ok-button">
          Ok
        </Button>
      </DialogActions>
    </Dialog>
  );
}
