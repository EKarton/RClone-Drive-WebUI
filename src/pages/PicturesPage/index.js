import { useState } from 'react';
import { useNavigate } from 'react-router';
import RemotesListSection from 'components/RemotesListSection';
import { FileViewerDialogProvider } from 'contexts/FileViewerDialog/index';
import { hashRemotePath } from 'utils/remote-paths-url';
import FolderBrowserDialog from './FolderBrowserDialog';
import RecentPicturesSection from './RecentPicturesSection';
import './index.scss';

/**
 * Represents the Pictures page with the list of remotes
 */
export default function PicturesPage() {
  const navigate = useNavigate();

  const [selectedRemote, setSelectedRemote] = useState();
  const [isDialogOpen, setIsDialogOpen] = useState(false);

  const handleRemoteCardClicked = (remote) => {
    setSelectedRemote(remote);
    setIsDialogOpen(true);
  };

  const renderFolderDialogTitle = () => (
    <>
      Select which folder to view pictures on <strong>{selectedRemote}</strong>
    </>
  );

  const handleFolderDialogCancelled = () => {
    setIsDialogOpen(false);
  };

  const handleFolderDialogSelected = (remotePath) => {
    navigate(`/pictures/${hashRemotePath(remotePath)}`);
    setIsDialogOpen(false);
  };

  return (
    <FileViewerDialogProvider>
      <div className="pictures-page">
        <FolderBrowserDialog
          title={renderFolderDialogTitle()}
          remotes={[selectedRemote]}
          open={isDialogOpen}
          onCancel={handleFolderDialogCancelled}
          onOk={handleFolderDialogSelected}
        />
        <RecentPicturesSection />
        <RemotesListSection onRemoteCardClicked={handleRemoteCardClicked} />
      </div>
    </FileViewerDialogProvider>
  );
}
