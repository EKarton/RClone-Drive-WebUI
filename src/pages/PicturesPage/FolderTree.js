import TreeView from "@mui/lab/TreeView";
import TreeItem from "@mui/lab/TreeItem";
import ExpandMoreIcon from "@mui/icons-material/ExpandMore";
import ChevronRightIcon from "@mui/icons-material/ChevronRight";
import useRCloneClient from "hooks/useRCloneClient";
import { useEffect, useState } from "react";
import { StatusTypes } from "pages/PicturesListPage";

const FolderTree = ({ remotes, onFolderSelect }) => {
  const handleNodeSelect = (_e, nodeIds) => {
    onFolderSelect(nodeIds);
  };

  return (
    <TreeView
      aria-label="file tree navigator"
      defaultCollapseIcon={<ExpandMoreIcon />}
      defaultExpandIcon={<ChevronRightIcon />}
      onNodeSelect={handleNodeSelect}
    >
      {remotes.map((remote) => (
        <FolderTreeItem remote={remote} curPath="" label={remote} />
      ))}
    </TreeView>
  );
};

const FolderTreeItem = ({ remote, curPath, label }) => {
  const rCloneClient = useRCloneClient();

  const [status, setStatus] = useState(StatusTypes.LOADING);
  const [subFolders, setSubFolders] = useState(undefined);
  const [error, setError] = useState(null);

  const [isExpanded, setIsExpanded] = useState(false);

  useEffect(() => {
    const fetchFolders = async () => {
      try {
        setError(null);
        setStatus(StatusTypes.LOADING);

        const data = await rCloneClient.fetchSubFolders(remote, curPath);

        setSubFolders(data.sort());
        setStatus(StatusTypes.SUCCESS);
      } catch (err) {
        setError(err);
      }
    };

    if (isExpanded) {
      fetchFolders();
    }
  }, [curPath, isExpanded, rCloneClient, remote]);

  const handleTreeItemClicked = () => {
    setIsExpanded(true);
  };

  const renderSubFolders = () => {
    if (!isExpanded) {
      return <div />;
    }

    if (status === StatusTypes.LOADING) {
      return <div>Loading</div>;
    }

    if (status === StatusTypes.ERROR) {
      return <div>Error! {error}</div>;
    }

    if (subFolders.length === 0) {
      return <div>No subfolders</div>;
    }

    return subFolders.map((subFolder) => (
      <FolderTreeItem
        remote={remote}
        curPath={subFolder.Path}
        label={subFolder.Name}
      />
    ));
  };

  return (
    <TreeItem
      nodeId={`${remote}:${curPath}`}
      label={label}
      onClick={handleTreeItemClicked}
      expandIcon={<ChevronRightIcon />}
      collapseIcon={<ExpandMoreIcon />}
    >
      {renderSubFolders()}
    </TreeItem>
  );
};

export default FolderTree;
