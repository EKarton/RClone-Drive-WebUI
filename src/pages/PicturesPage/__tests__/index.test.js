import RemotesListSection from 'components/RemotesListSection';
import { FileViewerDialogProvider } from 'contexts/FileViewerDialog';
import PicturesPage from 'pages/PicturesPage';
import { hashRemotePath } from 'utils/remote-paths-url';
import { act, customRender, waitFor } from 'test-utils/react';
import FolderBrowserDialog from '../FolderBrowserDialog';
import RecentPicturesSection from '../RecentPicturesSection';

jest.mock('components/RemotesListSection');
jest.mock('../FolderBrowserDialog');
jest.mock('../RecentPicturesSection');

describe('PicturesPage', () => {
  beforeEach(() => {
    FolderBrowserDialog.mockReturnValue(null);
    RecentPicturesSection.mockReturnValue(null);
    RemotesListSection.mockReturnValue(null);
  });

  it('should open FolderBrowserDialog when user clicks on a remote', async () => {
    // Mock user clicking on a remote
    jest.useFakeTimers();

    RemotesListSection.mockImplementation(({ onRemoteCardClicked }) => {
      setTimeout(() => onRemoteCardClicked('googledrive'), 10000);
      return null;
    });

    renderComponent();

    act(() => jest.runAllTimers());

    await waitFor(() => {
      expect(FolderBrowserDialog.mock.calls[1][0].open).toBeTruthy();
    });
  });

  it('should redirect to correct page when user selected a remote in the FolderBrowserDialog component', async () => {
    jest.useFakeTimers();

    FolderBrowserDialog.mockImplementationOnce(({ onOk }) => {
      setTimeout(() => onOk('googledrive:Pictures'), 10000);
      return null;
    });

    const view = renderComponent();

    act(() => jest.runAllTimers());

    await waitFor(() => {
      const expectedPath = `/pictures/${hashRemotePath('googledrive:Pictures')}`;
      expect(view.history.location.pathname).toEqual(expectedPath);
    });
  });

  it('should close the dialog when user selects nothing in the FolderBrowserDialog component', async () => {
    jest.useFakeTimers();

    FolderBrowserDialog.mockImplementationOnce(({ onOk }) => {
      setTimeout(() => onOk(''), 10000);
      return null;
    });

    renderComponent();

    act(() => jest.runAllTimers());

    await waitFor(() => {
      expect(FolderBrowserDialog.mock.calls[0][0].open).toBeFalsy();
    });
  });

  it('should close the dialog when user closes the dialog', async () => {
    jest.useFakeTimers();

    FolderBrowserDialog.mockImplementationOnce(({ onCancel }) => {
      setTimeout(() => onCancel(), 10000);
      return null;
    });

    renderComponent();

    act(() => jest.runAllTimers());

    await waitFor(() => {
      expect(FolderBrowserDialog.mock.calls[0][0].open).toBeFalsy();
    });
  });

  const renderComponent = () => {
    const initialRCloneInfoState = {
      endpoint: 'http://localhost:5572',
      username: 'local',
      password: '1234',
    };

    const componentToRender = (
      <FileViewerDialogProvider>
        <PicturesPage />
      </FileViewerDialogProvider>
    );

    const view = customRender(componentToRender, {
      initialRCloneInfoState,
    });

    return view;
  };
});
