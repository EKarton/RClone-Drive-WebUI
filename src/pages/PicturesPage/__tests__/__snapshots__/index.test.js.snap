// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PicturesPage should match snapshot when api call succeeds 1`] = `
<body>
  <div />
</body>
`;

exports[`PicturesPage should match snapshot when api call succeeds 2`] = `
<body>
  <div>
    <div
      class="filespage"
    >
      <div>
        <button
          class="MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButtonBase-root css-1rwt2y5-MuiButtonBase-root-MuiButton-root"
          data-testid="googledrive"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiButton-startIcon MuiButton-iconSizeMedium css-1d6wzja-MuiButton-startIcon"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="StorageIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M2 20h20v-4H2v4zm2-3h2v2H4v-2zM2 4v4h20V4H2zm4 3H4V5h2v2zm-4 7h20v-4H2v4zm2-3h2v2H4v-2z"
              />
            </svg>
          </span>
          googledrive
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </div>
      <div>
        <button
          class="MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButtonBase-root css-1rwt2y5-MuiButtonBase-root-MuiButton-root"
          data-testid="icloud"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiButton-startIcon MuiButton-iconSizeMedium css-1d6wzja-MuiButton-startIcon"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="StorageIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M2 20h20v-4H2v4zm2-3h2v2H4v-2zM2 4v4h20V4H2zm4 3H4V5h2v2zm-4 7h20v-4H2v4zm2-3h2v2H4v-2z"
              />
            </svg>
          </span>
          icloud
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </div>
      <div>
        <button
          class="MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButtonBase-root css-1rwt2y5-MuiButtonBase-root-MuiButton-root"
          data-testid="onedrive"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiButton-startIcon MuiButton-iconSizeMedium css-1d6wzja-MuiButton-startIcon"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="StorageIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M2 20h20v-4H2v4zm2-3h2v2H4v-2zM2 4v4h20V4H2zm4 3H4V5h2v2zm-4 7h20v-4H2v4zm2-3h2v2H4v-2z"
              />
            </svg>
          </span>
          onedrive
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
  </div>
</body>
`;
